app/v1/asset/keeper.go

```diff
@@ -18,24 +21,200 @@ type Keeper struct {
+		"moniker", []byte(msg.Moniker),
+	)
+
+	return createTags, nil
+}
+
+// EditGateway edits the specified gateway
+func (k Keeper) EditGateway(ctx sdk.Context, msg MsgEditGateway) (sdk.Tags, sdk.Error) {
+	// TODO
+	return nil, nil
+}
+
+// GetGateway retrieves the gateway of the given id
+func (k Keeper) GetGateway(ctx sdk.Context, gatewayID uint64) (Gateway, sdk.Error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(KeyGateway(gatewayID))
+	if bz == nil {
+		return Gateway{}, ErrUnkwownGateway(k.codespace, fmt.Sprintf("Unknown gateway id:%d", gatewayID))
+	}
+
+	var gateway Gateway
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &gateway)
+
+	return gateway, nil
+}
+
+// GetGatewayByMoniker retrieves the gateway of the given moniker
+func (k Keeper) GetGatewayByMoniker(ctx sdk.Context, moniker string) (Gateway, sdk.Error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(KeyMoniker(moniker))
+	if bz == nil {
+		return Gateway{}, ErrUnkwownGateway(k.codespace, fmt.Sprintf("Unknown gateway moniker:%s", moniker))
+	}
+
+	var gatewayID uint64
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &gatewayID)
+
+	return k.GetGateway(ctx, gatewayID)
+}
+
+// HasGateway checks if the given gateway exists. Return true if exists, false otherwise
+func (k Keeper) HasGateway(ctx sdk.Context, moniker string) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(KeyMoniker(moniker))
+}
+
+// saveGateway saves the gateway, which behaves by the specified mode
+func (k Keeper) saveGateway(ctx sdk.Context, gateway Gateway, creation bool) {
+	k.setGateway(ctx, gateway)
+
+	if creation {
+		k.setMoniker(ctx, gateway.Moniker, gateway.ID)
+		k.setOwnerGatewayID(ctx, gateway.Owner, gateway.ID)
+		k.increaseOwnerGatewayCount(ctx, gateway.Owner)
+		k.increaseGatewayCount(ctx)
+	}
+}
+
+// setGateway stores the given gateway into underlying storage
+func (k Keeper) setGateway(ctx sdk.Context, gateway Gateway) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gateway)
+
+	// set KeyGateway
+	store.Set(KeyGateway(gateway.ID), bz)
+}
+
+// setMoniker stores the gateway ID into storage by the key KeyMoniker
+func (k Keeper) setMoniker(ctx sdk.Context, moniker string, gatewayID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gatewayID)
+
+	// set KeyMoniker
+	store.Set(KeyMoniker(moniker), bz)
+}
+
+// setOwnerGatewayID stores the gateway ID into storage by the key KeyOwnerGateway. Intended for iteration on ids of an owner
+func (k Keeper) setOwnerGatewayID(ctx sdk.Context, owner sdk.AccAddress, gatewayID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gatewayID)
+
+	// set KeyOwnerGatewayID
+	store.Set(KeyOwnerGatewayID(owner, gatewayID), bz)
+}
+
+// increaseOwnerGatewayCount increases the count of the gateways of the given owner
+func (k Keeper) increaseOwnerGatewayCount(ctx sdk.Context, owner sdk.AccAddress) {
```

Comment:
I do not think it is necessary to store the count on blockchain side, it should be done in client apps, what do you think @secret2830 @chengwenxi

Reply:
Agree with you.

Reply:
Reserved for the later consideration?  Assume that we allow a gateway owner to query the gateway count via CLI, then this will be needed.